File: scala/src/test/scala/ModifyingSequenceSpec.scala
import org.scalatest._
import collection.mutable.ArrayBuffer

import algorithm._

class ModifyingSequenceSpec extends FlatSpec with Matchers {

  def isEven(x: Int) : Boolean = x % 2 == 0
  def beOne() : Int = 1

  trait Fixture {
    var a = ArrayBuffer(1, 1)
    var b = ArrayBuffer(2, 3)
    val empty : ArrayBuffer[Int] = ArrayBuffer()
  }

  "Modifying Sequence Algorithms" should "Copy OK" in new Fixture {
    ModifyingSequence.Copy(a, b) should be(ArrayBuffer(2, 3))
  }

  it should "CopyIf OK" in new Fixture {
    ModifyingSequence.CopyIf(a, b, isEven) should be(ArrayBuffer(2, 1))
    ModifyingSequence.CopyIf(a, b, (x: Int) => false) should be(a)
    ModifyingSequence.CopyIf(a, empty, isEven) should be(a)
  }

  it should "CopyN OK" in new Fixture {
    ModifyingSequence.CopyN(a, empty, 2) should be(ArrayBuffer(1, 1))
    ModifyingSequence.CopyN(a, b, 1) should be(ArrayBuffer(2, 1))
    ModifyingSequence.CopyN(a, b, 3) should be(ArrayBuffer(2, 3))
  }

  it should "CopyBackward OK" in new Fixture {
    val short = ArrayBuffer(2)
    val long = ArrayBuffer(1, 2, 3)
    ModifyingSequence.CopyBackward(a, short) should be(ArrayBuffer(1, 2))
    ModifyingSequence.CopyBackward(a, long) should be(ArrayBuffer(2, 3))
  }

  it should "Fill OK" in new Fixture {
    ModifyingSequence.Fill(empty, 3) should be(empty)
    ModifyingSequence.Fill(a, 3) should be(ArrayBuffer(3, 3))
  }

  it should "FillN OK" in new Fixture {
    ModifyingSequence.FillN(empty, 2, 1) should be(empty)
    ModifyingSequence.FillN(a, 3, 2) should be(ArrayBuffer(2, 2))
    ModifyingSequence.FillN(a, 1, 2) should be(ArrayBuffer(2, 1))
  }

  it should "Transform OK" in new Fixture {
    def plusOne = (x: Int) => x + 1
    ModifyingSequence.Transform(a, plusOne) should be(ArrayBuffer(2, 2))
  }

  it should "Generate OK" in new Fixture {
    ModifyingSequence.Generate(b, beOne) 
    b should be(ArrayBuffer(1, 1))
  }

  it should "GenerateN OK" in new Fixture {
    val seq = ModifyingSequence.GenerateN(3, beOne) 
    seq should contain only(1)
    seq should have size 3
  }

  it should "Remove OK" in new Fixture {
    ModifyingSequence.Remove(b, 2) should be(ArrayBuffer(3))
  }

  it should "RemoveIf OK" in new Fixture {
    ModifyingSequence.RemoveIf(b, isEven) should be(ArrayBuffer(3))
  }

  it should "RemoveCopy OK" in new Fixture {
    val (removed, remained) = ModifyingSequence.RemoveCopy(b, 2)
    removed should be(ArrayBuffer(2))
    remained should be(ArrayBuffer(3))
  }

  it should "RemoveCopyIf OK" in new Fixture {
    val (removed, remained) = ModifyingSequence.RemoveCopyIf(b, isEven)
    removed should be(ArrayBuffer(2))
    remained should be(ArrayBuffer(3))
  }

  it should "Replace OK" in new Fixture {
    ModifyingSequence.Replace(a, 1, 2)
    a should be(ArrayBuffer(2, 2))
  }

  it should "ReplaceIf Ok" in new Fixture {
    ModifyingSequence.ReplaceIf(b, isEven, 4)
    b should be(ArrayBuffer(4, 3))
  }

  it should "ReplaceCopy OK" in new Fixture {
    ModifyingSequence.ReplaceCopy(a, 1, 2) should be(ArrayBuffer(2, 2))
  }

  it should "ReplaceCopyIf Ok" in new Fixture {
    ModifyingSequence.ReplaceCopyIf(b, isEven, 4) should be(ArrayBuffer(4, 3))
  }

  it should "RangeSwap OK" in new Fixture {
    ModifyingSequence.RangeSwap(a, b)
    a should be (ArrayBuffer(2, 3))
    b should be(ArrayBuffer(1, 1))

    var c = ArrayBuffer(3, 4)
    var short = ArrayBuffer(1)
    ModifyingSequence.RangeSwap(c, short)
    c should be(ArrayBuffer(1, 4))
    short should be(ArrayBuffer(3))
  }

  it should "Reverse OK" in new Fixture {
    ModifyingSequence.Reverse(b)
    b should be(ArrayBuffer(3, 2))
  }

  it should "ReverseCopy OK" in new Fixture {
    ModifyingSequence.ReverseCopy(b) should be(ArrayBuffer(3, 2))
  }
}

File: typescript/test/modifyingSequenceSpec.js
var should = require('chai').should();
var Algos = require('../dist/modifyingSequence');

var Ary = function() {
  return [1, 2, 3, 4, 5];
};

var Hash = function() {
  return {'a': 1, 'b': 2};
};

describe('Modifying sequence algorithms', function(){

  it('copy', function(){
    var hash = {};
    Algos.copy(Hash(), hash);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1, 'b': 2}));
  });

  it('copyIf', function(){
    var hash = {};
    Algos.copyIf(Hash(), hash, function(e){
      return e === 1;
    });
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1}));
  });

  it('copyN', function(){
    var ary = new Array(3);
    Algos.copyN(ary, 3, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, 3]));
  });

  it('copyBackward', function(){
    var ary = new Array(3);
    Algos.copyBackward(Ary(), ary).should.be.equal(-1);
    ary = new Array(5);
    Algos.copyBackward(Ary(), ary).should.be.equal(0);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2, 3, 4, 5]));
    ary = new Array(6);
    ary[0] = 1;
    Algos.copyBackward(Ary(), ary).should.be.equal(1);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 1, 2, 3, 4, 5]));
  });

  it('fill', function(){
    var ary = new Array(3);
    Algos.fill(ary, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, 3]));
  });

  it('fillN', function(){
    var ary = new Array(3);
    Algos.fillN(ary, 2, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, null]));
  });

  it('transform', function(){
    var ary = Ary()
    JSON.stringify(Algos.transform(ary, function(e){
      return e + 1
    }));
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 3, 4, 5, 6]));
  });

  it('generate', function(){
    var ary = new Array(3);
    Algos.generate(ary, function(){
      return 3;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, 3]));
  });

  it('generateN', function(){
    var ary = new Array(3);
    Algos.generateN(ary, 2, function(){
      return 3;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, null]));
  });

  it('remove', function(){
    var ary = [1, 2, 3];
    Algos.remove(ary, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2]));
  });

  it('removeIf', function(){
    var ary = [1, 2, 3];
    Algos.removeIf(ary, function(e){
      return e < 2;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 3]));
    var hash = {'a': 1, 'b': 2};
    Algos.removeIf(hash, function(e){
      return e != 1;
    });
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1}));
  });

  it('removeCopy', function(){
    var ary = [1, 2, 3];
    JSON.stringify(Algos.removeCopy(ary, 3))
      .should.be.equal(JSON.stringify([3]));
  });

  it('removeCopyIf', function(){
    var ary = [1, 2, 3];
    ary = Algos.removeCopyIf(ary, function(e){
      return e < 2;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([1]));
    var hash = {'a': 1, 'b': 2};
    hash = Algos.removeCopyIf(hash, function(e){
      return e != 1;
    });
    JSON.stringify(hash).should.be.equal(JSON.stringify({'b': 2}));
  });

  it('replace', function(){
    var ary = [1, 2, 3, 3];
    Algos.replace(ary, 3, 2);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2, 2, 2]));
  });

  it('replaceIf', function(){
    var ary = [1, 2, 3];
    Algos.replaceIf(ary, function(e){
      return e < 2;
    }, 2);
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 2, 3]));
    var hash = {'a': 1, 'b': 2};
    Algos.replaceIf(hash, function(e){
      return e != 2;
    }, 2);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 2, 'b': 2}));
  });

  it('replaceCopy', function(){
    var ary = [1, 2, 3, 3];
    JSON.stringify(Algos.replaceCopy(ary, 3, 2))
      .should.be.equal(JSON.stringify([1, 2, 2, 2]));
  });

  it('replaceCopyIf', function(){
    var ary = [1, 2, 3];
    ary = Algos.replaceCopyIf(ary, function(e){
      return e < 2;
    }, 2);
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 2, 3]));
    var hash = {'a': 1, 'b': 2};
    hash = Algos.replaceCopyIf(hash, function(e){
      return e != 2;
    }, 2);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 2, 'b': 2}));
  });

  it('swapRange', function(){
    var a = [1, 2, 3, 4];
    var b = [4, 3, 2, 1];
    Algos.swapRange(a, b, 1, 3);
    JSON.stringify(a).should.be.equal(JSON.stringify([1, 3, 2, 4]));
    JSON.stringify(b).should.be.equal(JSON.stringify([4, 2, 3, 1]));
  });

  it('reverse', function(){
    var ary = [1, 2, 3];
    Algos.reverse(ary);
    ary.toString().should.be.equal([3, 2, 1].toString());
  });

  it('reverseCopy', function(){
    Algos.reverseCopy(Ary()).toString().
      should.be.equal(Ary().reverse().toString());
  });

  it('shuffle', function(){
    Algos.shuffle(Ary());
  });

  it('rotate', function(){
    var ary = [1, 2, 3, 4];
    Algos.rotate(ary, 2);
    ary.toString().should.be.equal([3, 4, 1, 2].toString());
  });

  it('rotateCopy', function(){
    Algos.rotateCopy(Ary(), 2).toString()
      .should.be.equal([3, 4, 5, 1, 2].toString());
  });

  it('unique', function(){
    var duplicate = [1, 1, 2, 3, 3];
    Algos.unique(duplicate);
    JSON.stringify(duplicate).should.be.equal(JSON.stringify([1, 2, 3]));
  });

  it('uniqueCopy', function(){
    var duplicate = [1, 1, 2, 3, 3];
    JSON.stringify(Algos.uniqueCopy(duplicate))
      .should.be.equal(JSON.stringify([1, 2, 3]));
  });
});
